### ISSUE TITLE
Help wanted: shell scripting

### ISSUE THREAD

**Issue by barrydegraaff** [Zimbra-Community/owncloud-zimlet#91]
Anyone here who has knows shell scripting and willing to put some time 
in the community? Let me know!

I am looking for someone who can add an upgrade feature to the WebDAV 
Client installer, aka an upgrade feature. Right not the installer
only fresh installs.

Your code needs to parse config_template.xml and config.properties, 
read the current values, compare any changes, and put them pack after upgrading:
https://github.com/Zimbra-Community/owncloud-zimlet/blob/soapServiceBarry/webdav-client-installer.sh

Also, your work needs to be re-usable to it can be used for other Zimlet,
there is no $$$ available for this, its a voluntary effort.

If you are interested, let me know!

**Comment by Dual-Boot** [comment]
Hi Barry,

just see you open that thread. Maybe it is easyer for you that I post my sample here.
I have push the check about mailboxd service into the file you will get into attachment.

 I need some serious learning about Git.

```#!/bin/bash

# Copyright (C) 2016  Barry de Graaff
# 
# Bugs and feedback: https://github.com/Zimbra-Community/owncloud-zimlet/issues
# 
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 2 of the License, or
# (at your option) any later version.
# 
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
# 
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see http://www.gnu.org/licenses/.

set -o errexit
set -o nounset
# if you want to trace your script uncomment the following line
#set -o xtrace

echo "Automated Zimbra WebDAV Client installer for single-server Zimbra 8.7 on CentOS 6 or 7 (Ubuntu untested)
- Installs ant and git, the WebDAV Client server extension and Zimlet."

# Make sure only root can run our script
if [ "$(id -u)" != "0" ]; then
   echo "This script must be run as root" 1>&2
   exit 1
fi

# check if a service mailboxd is running on this server
MAILBOXD="$(su - zimbra << 'EOF'
        zmcontrol status | grep mailboxd || true
EOF
)"
if [ ! -z "${MAILBOXD}" ] ;
        then echo "mailboxd is running"
        else echo "mailboxd is not present on this server"
        exit 1
fi

echo ""
echo "Do you want to enable experimental document preview (CentOS7 only)? Y/n:"
read YN;

echo "Check if git and ant are installed."
YUM_CMD=$(which yum || true)
APT_CMD=$(which apt-get || true)
GIT_CMD=$(which git || true)
ANT_CMD=$(which ant || true)
ZIP_CMD=$(which zip || true)

if [[ -z $GIT_CMD ]] || [[ -z $ANT_CMD ]] || [[ -z $ZIP_CMD ]]; then
   if [[ ! -z $YUM_CMD ]]; then
      yum install -y git ant make zip
   else
      apt-get install -y git ant make default-jdk zip
   fi
fi


echo "Remove old versions of Zimlet."
rm -Rf /opt/zimbra/zimlets-deployed/_dev/tk_barrydegraaff_owncloud_zimlet/
su - zimbra -c "zmzimletctl undeploy tk_barrydegraaff_owncloud_zimlet"


TMPFOLDER="$(mktemp -d /tmp/webdav-client-installer.XXXXXXXX)"
echo "Download WebDAV Client to ${TMPFOLDER}"
cd ${TMPFOLDER}
git clone --depth=1 https://github.com/barrydegraaff/owncloud-zimlet
#cp -r /root/owncloud-zimlet ${TMPFOLDER}

echo "Compiling WebDAV Client"
cd owncloud-zimlet
cd extension && ant download-libs && cd ..
make


echo "Installing server extension to /opt/zimbra/lib/ext/ownCloud"
cd ${TMPFOLDER}/owncloud-zimlet/dist/owncloud-extension/
shopt -s extglob
ZAL_VERSION="1.11"
ZAL_VERSION_EXTENDED="1.11.2"
ZIMBRA_VERSION=$(sudo su - zimbra -c "zmcontrol -v" | tr -d '\n' | sed -r 's/.* ([0-9\.]+[0-9]).*/\1/')
echo "Downloading the correct ZAL Version (${ZAL_VERSION_EXTENDED} for zimbra ${ZIMBRA_VERSION})..."
wget -c "https://openzal.org/${ZAL_VERSION}/zal-${ZAL_VERSION_EXTENDED}-${ZIMBRA_VERSION}.jar" -O "zal-${ZAL_VERSION_EXTENDED}-${ZIMBRA_VERSION}.jar"
mkdir -p /opt/zimbra/lib/ext/ownCloud
rm -f /opt/zimbra/lib/ext/ownCloud/*.jar
cp "zal-${ZAL_VERSION_EXTENDED}-${ZIMBRA_VERSION}.jar" /opt/zimbra/lib/ext/ownCloud/
cp !(zal*).jar /opt/zimbra/lib/ext/ownCloud/

echo "allowdomains=*" > /opt/zimbra/lib/ext/ownCloud/config.properties

ls -hal /opt/zimbra/lib/ext/ownCloud/

echo "Installing Zimlet"
mkdir -p /opt/zimbra/zimlets-deployed/_dev/tk_barrydegraaff_owncloud_zimlet/
unzip ${TMPFOLDER}/owncloud-zimlet/zimlet/tk_barrydegraaff_owncloud_zimlet.zip -d /opt/zimbra/zimlets-deployed/_dev/tk_barrydegraaff_owncloud_zimlet/

echo "Flushing Zimlet Cache"
su - zimbra -c "zmprov fc all"

if [ "$YN" == 'Y' ];
then
   echo "Install LibreOffice"
   cp -v ${TMPFOLDER}/owncloud-zimlet/bin/* /usr/local/sbin/
   yum install -y libreoffice-headless.x86_64 libreoffice.x86_64
   
   echo "Configure docconvert user and set up sudo in /etc/sudoers.d/99_zimbra-docconvert"
   adduser docconvert || true
   echo "zimbra     ALL=(docconvert) NOPASSWD: ALL" > /etc/sudoers.d/99_zimbra-docconvert
   usermod -a -G zimbra docconvert
   usermod -a -G docconvert zimbra
fi

echo "--------------------------------------------------------------------------------------------------------------
Zimbra WebDAV Client installed successful

To load the extension:

su zimbra
zmmailboxdctl restart

  As of version 0.5.6 your clients CAN CONNECT TO ALL DAV SERVERS BY DEFAULT,
  you can restrict the allowed DAV servers to connect to in:

  /opt/zimbra/lib/ext/ownCloud/config.properties
  allowdomains=allowme.example.com;allowmealso.example.com

  - No service restart is needed after changing this file.

  If you installed WebDAV Client before, you should remove your DAV servers
  from zimbraProxyAllowedDomains:
  zmprov gc default zimbraProxyAllowedDomains
  zmprov mc default -zimbraProxyAllowedDomains allowme.example.com

"

rm -Rf ${TMPFOLDER}

exit 0```

**Comment by Dual-Boot** [comment]
Hello Barry,

here is a code sample that you can add to check if mailboxd is running where you execute the installer : 

```
MAILBOXD="$(su - zimbra << 'EOF'
        source ~/bin/zmshutil ;
        zmsetvars ;
        ldapsearch -LLL -x -H $ldap_master_url -D $zimbra_ldap_userdn -w $zimbra_ldap_password \
                -b "cn=$(zmhostname),cn=servers,cn=zimbra" \
                zimbraServiceEnabled=mailbox zimbraServiceEnabled | grep mailbox || true
EOF
)"

if [ ! -z "${MAILBOXD}" ] ;
        then echo "mailboxd is running"
        else echo "mailboxd is not present on this server"
        exit 1
fi
```

**Comment by barrydegraaff** [comment]
@Dual-Boot thanks, however your code does not seem to work in all cases, e.g. when ldap is down:

    ldap_sasl_bind(SIMPLE): Can't contact LDAP server (-1)
    Last login: Mon Nov 7 07:22:18 CET 2016
    mailboxd is enabled on this server

Also, I need tested code... otherwise I might as well do it myself.

**Comment by barrydegraaff** [comment]
besides, why not use zmprov gcf or something? Also this issue is about upgrading .properties and config_template...

**Comment by barrydegraaff** [comment]
part one:
https://github.com/Zimbra-Community/propmigr

**Comment by barrydegraaff** [comment]
and this:
https://github.com/Zimbra-Community/prop2xml

**Comment by barrydegraaff** [comment]
fixed in 0.6.5

**Comment by Dual-Boot** [comment]
Hi Barry,

 my point of view was to not rely on java process to have a fast script.
You're right when talk about LDAP server. When LDAP server is down the piece of code
does not work. But If LDAP server is down the good question is : do I need to install
or upgrade the Zimlet ? The answer is : No . First I need a clean environment before
installing any Zimlet.

 Which leads me to a point : I need different levels of checks.

- my mail service is a mono or multi-servers installation (not mandatory I think)
- how many number of LDAP servers
- which servers are Masters
- which LDAP server do I need to ask to know if my server is running or not a mailbox
- If a Master does not answer what is the best way to act ? Stopping the installation process
or asking a second Master ?

 Regards,


 Regards,

----- Mail original -----
> De: "Barry de Graaff" <notifications@github.com>
> À: "Zimbra-Community/owncloud-zimlet" <owncloud-zimlet@noreply.github.com>
> Cc: "DualBoot" <guycarre@free.fr>, "Mention" <mention@noreply.github.com>
> Envoyé: Lundi 7 Novembre 2016 07:26:50
> Objet: Re: [Zimbra-Community/owncloud-zimlet] Help wanted: shell scripting (#91)
> 
> 
> 
> @Dual-Boot thanks, however your code does not seem to work in all
> cases, e.g. when ldap is down: ldap_sasl_bind(SIMPLE): Can't contact
> LDAP server (-1)
> Last login: Mon Nov 7 07:22:18 CET 2016
> mailboxd is enabled on this server
> 
> Also, I need tested code... otherwise I might as well do it myself.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub , or mute the thread
> .
> 
> 

-- 
  ___________________________________
 /                                   \
/-------------------------------------\
|  Guy CARRÉ                          |
|  *************                      |
|  PostMaster - WikiMaster - SysAdmin |
|                                     |
| "Free Your Mind. Think Open Source" |
|  april.org                          |
|                                     |
|_____________________________________|
